File: Loom.Engine/Assemblers/PromptAssembler.vb
Imports System.Text
Imports Loom.Core.Enums
Imports Loom.Core.Interfaces
Imports Loom.Core.Models

Namespace Assemblers
    ''' <summary>
    ''' Assembles the final list of messages to be sent to the LLM provider.
    ''' Combines the original conversation messages with the invocation's contextual information by injecting the generated
    ''' context block into the system message or creating one if missing
    ''' </summary>
    Public Class PromptAssembler
        Implements IAssembler
        ''' <summary>
        ''' Entry point: transform invocation into a messages list
        ''' </summary>
        ''' <param name="invocation"></param>
        ''' <returns> List of all messages </returns>
        Public Function Assemble(invocation As LlmInvocation) As List(Of Message) Implements IAssembler.Assemble

            ' Copy of original messages list
            Dim finalMessages As New List(Of Message)(invocation.Conversation.Messages)

            Dim contextBlock = BuildContextBlock(invocation)

            ' Finds system message or creates a new one on top
            Dim systemMsg = finalMessages.FirstOrDefault(Function(m) m.Role = MessageRole.System)

            If systemMsg IsNot Nothing Then

                systemMsg.Content = AddContent(systemMsg.Content, contextBlock)

            Else

                finalMessages.Insert(0, New Message With {
                    .Role = MessageRole.System,
                    .Content = AddContent(invocation.Conversation.SystemPrompt, contextBlock)
                })

            End If

            Return finalMessages

        End Function

        ''' <summary>
        ''' Assembles each role kind with its content (user/tool/assistant)
        ''' </summary>
        ''' <param name="invocation"></param>
        ''' <returns> Object with planned structure </returns>
        Public Function AssembleInputItems(invocation As LlmInvocation) As List(Of AssembledItem) Implements IAssembler.AssembleInputItems
            Dim msgs = Me.Assemble(invocation)
            Dim items As New List(Of AssembledItem)

            For Each msg In msgs
                If msg.Role = MessageRole.Tool Then
                    items.Add(New AssembledItem With {
                        .Role = "assistant",
                        .Type = "function_call",
                        .ToolName = msg.ToolName,
                        .Content = msg.Content,
                        .ToolCallId = msg.ToolCallId,
                        .Arguments = msg.ToolArgs
                    })
                    items.Add(New AssembledItem With {
                        .Role = "tool",
                        .Type = "function_call_output",
                        .ToolCallId = msg.ToolCallId,
                        .ToolName = msg.ToolName,
                        .Content = msg.Content
                    })
                Else
                    items.Add(New AssembledItem With {
                        .Role = msg.Role.ToString().ToLower(),
                        .Type = "message",
                        .Content = msg.Content
                    })
                End If
            Next

            Return items
        End Function

        ''' <summary>
        ''' Builds the string that aggregates RAG & memory
        ''' </summary>
        ''' <param name="invocation"></param>
        ''' <returns> The complete memory and context prompt </returns>
        Private Function BuildContextBlock(invocation As LlmInvocation) As String Implements IAssembler.BuildContextBlock

            Dim sb As New StringBuilder()

            ' ############ Memory ############
            If Not String.IsNullOrWhiteSpace(invocation.Memory.Content) Then
                sb.AppendLine("### LONG-TERM MEMORY ###")
                sb.AppendLine(invocation.Memory.Content)
                sb.AppendLine()
            End If

            ' ############ RAG ############
            If invocation.Rag.RetrievedChunks.Any() Then

                sb.AppendLine("### CONTEXT ###")

                For Each chunk In invocation.Rag.RetrievedChunks
                    sb.AppendLine($"[Source: {chunk.SourceId}] - {chunk.Text}")
                Next
                sb.AppendLine()
            End If

            Return sb.ToString()

        End Function


        Private Function AddContent(original As String, context As String) As String Implements IAssembler.AddContent
            If String.IsNullOrWhiteSpace(context) Then Return original

            Return $"{original}{Environment.NewLine}{context}"

        End Function



    End Class
End Namespace

File: implementations/csharp/src/Loom.Engine/Assemblers/PromptAssembler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Loom.Core.Enums;
using Loom.Core.Interfaces;
using Loom.Core.Models;

namespace Loom.Engine.Assemblers
{
    public class PromptAssembler : IAssembler
    {
        /// <summary>
        /// Entry point: transform invocation into a messages list
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns> List of all messages </returns>
        public List<Message> Assemble(LlmInvocation invocation)
        {

            // Copy of original message list
            var finalMessages = new List<Message>(invocation.Conversation.Messages);

            var contextBlock = BuildContextBlock(invocation);

            // Finds system message or creates new one on top
            var systemMsg = finalMessages.FirstOrDefault(m => m.Role == MessageRole.System);

            if (systemMsg != null)
            {
                systemMsg.Content = AddContent(systemMsg.Content, contextBlock);
            } else
            {
                finalMessages.Insert(0, new Message
                {
                    Role = MessageRole.System,
                    Content = AddContent(invocation.Conversation.SystemPrompt, contextBlock)
                });
            }

            return finalMessages;
        }

        /// <summary>
        /// Assembles each role kind with its content (user/tool/assistant)
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns> Object with planned structure </returns>
        public List<AssembledItem> AssembleInputItems(LlmInvocation invocation)
        {

            var msgs = Assemble(invocation);
            var items = new List<AssembledItem>();

            foreach(var msg in msgs)
            {
                if (msg.Role == MessageRole.Tool)
                {
                    items.Add(new AssembledItem
                    {
                        Role = "assistant",
                        Type = "function_call",
                        ToolName = msg.ToolName,
                        Content = msg.Content,
                        ToolCallId = msg.ToolCallId,
                        Arguments = msg.ToolArgs
                    });
                    items.Add(new AssembledItem
                    {
                        Role = "tool",
                        Type = "function_call_output",
                        ToolCallId = msg.ToolCallId,
                        ToolName = msg.ToolName,
                        Content = msg.Content
                    });
                } else
                {
                    items.Add(new AssembledItem
                    {
                        Role = msg.Role.ToString().ToLower(),
                        Type = "message",
                        Content = msg.Content
                    });
                }
            }

            return items;
        }

        /// <summary>
        /// Builds the string that aggregates RAG and memory
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns> The complete memory and context prompt </returns>
        public string BuildContextBlock(LlmInvocation invocation)
        {

            var sb = new StringBuilder();

            // ############ Memory ############
            if (!string.IsNullOrWhiteSpace(invocation.Memory.Content))
            {
                sb.AppendLine("### LONG-TERM MEMORY ###");
                sb.AppendLine(invocation.Memory.Content);
                sb.AppendLine();
            }

            // ############ RAG ############
            if (invocation.Rag.RetrievedChunks.Any())
            {
                sb.AppendLine("### CONTEXT ###");

                foreach (var chunk in invocation.Rag.RetrievedChunks)
                {
                    sb.AppendLine($"[Source: {chunk.SourceId}] - {chunk.Text}");
                }
                sb.AppendLine();
            }

            return sb.ToString();
        }

        public string AddContent(string original, string context)
        {
            if (string.IsNullOrEmpty(context)) return original;

            return $"{original}{Environment.NewLine}{context}";
        }
    }
}
